File: problems_c++/QuickSort.cpp
#include <iostream>

using namespace std;

int partition(int array[],int f, int l){
	int aux;
	while(f<l){
		if(array[f]>array[l]){
			aux = array[f];
			array[f] = array[l-1];
			array[l-1] = array[l];
			array[l] = aux;
			l--;
		}else
			f++;
	}
	return l;
}

void quickSort(int array[], int f, int l){
	if(f < l){
		int p = partition(array,f,l);
		quickSort(array,f,p-1);
		quickSort(array,p+1,l);
	}
}

int main(){
	int array[10] = {5,7,2,8,9,10,1,85,0,12};
	quickSort(array,0,9);
	for(int i = 0;i<10;i++){
		cout<<array[i]<<" ";
	}
	cout<<endl;
	return 0;
}

File: technical_concepts_py/QuickSort.py
"""Implement quick sort in Python.
Input a list.
Output a sorted list."""

def partition(array, left, right, pivot):
	while left <= right:
		while array[left] < pivot:
			left += 1
		while array[right] > pivot:
			right -= 1
		if left <= right:
			tmp = array[left]
			array[left] = array[right]
			array[right] = tmp
			left += 1
			right -= 1
	return left 

def quicksortHelper(array, left, right):
	if left >= right:
		return

	pivot = array[(left + right) // 2]
	index = partition(array, left, right, pivot)
	quicksortHelper(array, left, index - 1)
	quicksortHelper(array, index, right)

def quicksort(array):
	if len(array) <= 1:
		return
	quicksortHelper(array, 0, len(array) - 1)
	
	

test = [5,7,2,8,9,10,1,85,0,12]
quicksort(test)
print(test)
